src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1375,41 +1376,65 @@ uct_dc_mlx5_iface_dci_do_rand_pending_tx(ucs_arbiter_t *arbiter,
     return res;
 }
 
-static ucs_arbiter_cb_result_t
-uct_dc_mlx5_ep_arbiter_purge_cb(ucs_arbiter_t *arbiter, ucs_arbiter_group_t *group,
-                                ucs_arbiter_elem_t *elem, void *arg)
+ucs_arbiter_cb_result_t
+uct_dc_mlx5_ep_arbiter_purge_internal_cb(ucs_arbiter_t *arbiter,
+                                         ucs_arbiter_group_t *group,
+                                         ucs_arbiter_elem_t *elem, void *arg)
 {
     uct_purge_cb_args_t *cb_args = arg;
     void **priv_args             = cb_args->arg;
     uct_dc_mlx5_ep_t *ep         = priv_args[0];
     uct_dc_mlx5_iface_t *iface   = ucs_derived_of(ep->super.super.iface,
                                                   uct_dc_mlx5_iface_t);
-    uct_pending_req_t *req       = ucs_container_of(elem, uct_pending_req_t, priv);
+    uct_pending_req_t *req       = ucs_container_of(elem, uct_pending_req_t,
+                                                    priv);
     uct_rc_pending_req_t *freq;
 
     if (uct_dc_mlx5_iface_is_dci_rand(iface) &&
         (uct_dc_mlx5_pending_req_priv(req)->ep != ep)) {
-        /* element belongs to another ep - do not remove it */
+        /* Element belongs to another ep - do not remove it */
         return UCS_ARBITER_CB_RESULT_NEXT_GROUP;
     }
 
-    if (ucs_likely(req->func != uct_dc_mlx5_iface_fc_grant)){
+    if (ucs_unlikely(req->func == uct_dc_mlx5_iface_fc_grant)) {
+        /* User callback should not be called for FC messages. Just return
+         * pending request memory to the pool */
+        freq = ucs_derived_of(req, uct_rc_pending_req_t);
+        ucs_mpool_put(freq);
+        return UCS_ARBITER_CB_RESULT_REMOVE_ELEM;
+    }
+
+    /* Non-internal request was found */
+    return UCS_ARBITER_CB_RESULT_RESCHED_GROUP;
+}
+
+static ucs_arbiter_cb_result_t
+uct_dc_mlx5_ep_arbiter_purge_cb(ucs_arbiter_t *arbiter, ucs_arbiter_group_t *group,
+                                ucs_arbiter_elem_t *elem, void *arg)
+{
+    uct_purge_cb_args_t *cb_args = arg;
+    void **priv_args             = cb_args->arg;
+    uct_dc_mlx5_ep_t *ep         = priv_args[0];
+    uct_pending_req_t *req       = ucs_container_of(elem, uct_pending_req_t,
+                                                    priv);
+    ucs_arbiter_cb_result_t result;
+
+    result = uct_dc_mlx5_ep_arbiter_purge_internal_cb(arbiter, group, elem,
+                                                      arg);
```

Comment:
you can pass ep pointer here instead of arg

Reply:
fixed